Describe more in README (#29)
konflux-ci/pipeline-migration-tool: README.md

Renovate is responsible for invoke migration tool properly. Then, in general, it is unnecessary for
Konflux users to run `migrate` by themselves.

To apply migrations, run command:

```bash
cd path/to/repo
pipeline-migration-tool migrate -u '<upgrades>'
```

`upgrades` is a JSON string encoded from a list of mappings. Each mapping includes data for a
single task bundle upgrade, for example:

```jsonc
[
  {
    "depName": "quay.io/konflux-ci/tekton-catalog/task-init",
    "currentValue": "0.1",
    "currentDigest": "sha256:...",
    "newValue": "0.1",
    "newDigest": "sha256:...",
    "packageFile": ".tekton/component-name-pull.yaml",
    "parentDir": ".tekton",
    "depTypes": ["tekton-bundle"]
  },
  // ...
]
```

The field names map to the [Renovate template fields]
(https://docs.renovatebot.com/templates/#other-available-fields) directly:

* `depName`: `{{depName}}`
* `currentValue`: `{{currentValue}}`
* `currentDigest`: `{{currentDigest}}`
* `newValue`: `{{newValue}}`
* `newDigest`: `{{newDigest}}`
* `packageFile`: `{{packageFile}}`
* `parentDir`: `{{parentDir}}`
* `depTypes`: `{{depTypes}}`

> To generate the list, handlebars built-in each helper of Renovate is used.

Reviewer: ```suggestion
To generate the list, handlebars built-in `each` helper of Renovate is used.
```

Without this the sentence may look confusing :)

A migration tool does migrations for Konflux pipelines.

Doing migrations is the major feature of pipeline-migration-tool. It auto-discovers migrations for
given task bundle upgrades and applies found migrations to build pipelines. The pipeline can be

> either included in a `PipelineRun` definition as `spec.pipelineSpec` or a single `Pipeline` defintion.

Reviewer 2: We need to say that it runs as a post upgrade task of Renovate.

Author: Updated.

`migrate` is the main one to do migrations. Following are a few examples of the commands usage. For
detailed information, refer to the help messages by running each one with `-h`.

### To apply migrations:

```bash
cd path/to/repo
pipeline-migration-tool migrate -u '<upgrades>'
```

`upgrades` is a JSON string encoded from a list of mappings. Each mapping includes data for a
single task bundle upgrade, for example:

```jsonc
[
  {
    "depName": "quay.io/konflux-ci/tekton-catalog/task-init",
    "currentValue": "0.1",
    "currentDigest": "sha256:...",
    "newValue": "0.1",
    "newDigest": "sha256:...",
    "packageFile": ".tekton/component-name-pull.yaml",
    "parentDir": ".tekton",
    "depTypes": ["tekton-bundle"]
  },
  // ...
]
```

The field names map to the [Renovate template fields]
(https://docs.renovatebot.com/templates/#other-available-fields) directly:

* `depName`: `{{depName}}`
* `currentValue`: `{{currentValue}}`
* `currentDigest`: `{{currentDigest}}`
* `newValue`: `{{newValue}}`
* `newDigest`: `{{newDigest}}`
* `packageFile`: `{{packageFile}}`
* `parentDir`: `{{parentDir}}`
* `depTypes`: `{{depTypes}}`

> To generate the lists, using handlebars built-in `each` helper supported by Renovate.

Reviewer 2: To generate the list, handlebars built-in `each` helper of Renovate is used.

### Add a Konflux task to build pipeline

Sub-command `add-task` provides rich options to add a Konflux task to build pipelines in local
Component repositories. Let's take the task `sast-coverity-check` as an example to see a few
command usages:

* Add task with latest bundle to pipelines from inside a repository:

  ```bash

>   pipeline-migration-tool add-task sast-coverity-check

Reviewer 2: It's not clear where the task is inserted.
Do we support any tasks or only from predefined list? It's not clear from this doc.

Author: > It's not clear where the task is inserted.

The next sentence tells that.

> Do we support any tasks or only from predefined list? It's not clear from this doc.

What do you mean "any tasks" and "predefined list"?

The first paragraph mentions that this sub-command adds a Konflux task to build pipelines. This README gives general description and `add-task -h` gives more detailed information.

Reviewer 2: > What do you mean "any tasks" and "predefined list"?

Which tasks a user can add? Can I do `pipeline-migration-tool add-task abcd1234`?

> This README gives general description and add-task -h gives more detailed information.

IMO, it would be better to have at least brief explanation here, so I don;t need to install the tool to know what it can do.

Author: > ### Add a Konflux task to build pipeline
> 
> Sub-command `add-task` provides rich options to add a Konflux task to build pipelines in local
> Component repositories. Let's take the task `sast-coverity-check` as an example to see a few
> command usages:

Both the heading and the paragraph mention that this is about adding a Konflux task. Isn't it enough to give enough context about what task can be added? Can you please suggest a description?

`migrate` is the main one to do migrations. There are a few examples below. For detailed
information, refer to the help messages by running each one with `-h`.

### To apply migrations:

```bash
cd path/to/repo
pipeline-migration-tool migrate -u '<upgrades>'
```

`upgrades` is a JSON string encoded from a list of mappings. Each mapping includes data for a
single task bundle upgrade, for example:

```jsonc
[
  {
    "depName": "quay.io/konflux-ci/tekton-catalog/task-init",
    "currentValue": "0.1",
    "currentDigest": "sha256:...",
    "newValue": "0.1",
    "newDigest": "sha256:...",
    "packageFile": ".tekton/component-name-pull.yaml",
    "parentDir": ".tekton",
    "depTypes": ["tekton-bundle"]
  },
  // ...
]
```

The field names map to the [Renovate template fields]
(https://docs.renovatebot.com/templates/#other-available-fields) directly:

* `depName`: `{{depName}}`
* `currentValue`: `{{currentValue}}`
* `currentDigest`: `{{currentDigest}}`
* `newValue`: `{{newValue}}`
* `newDigest`: `{{newDigest}}`
* `packageFile`: `{{packageFile}}`
* `parentDir`: `{{parentDir}}`
* `depTypes`: `{{depTypes}}`

To generate the list, handlebars built-in each helper of Renovate is used.

> pipeline-migration-tool is configured in Konflux Mintmaker as a Renovate post-upgrade command and

Reviewer 2: To me, it's crucial information which should be delivered to user at the beginning of the migrations paragraph. Also we can say, that besides the migrations, it can be useful in many other ways, like adding tasks, etc...

A migration tool does migrations for Konflux pipelines.

Doing migrations is the major feature of pipeline-migration-tool. It auto-discovers migrations for
given task bundle upgrades and applies found migrations to build pipelines. The pipeline can be
either included in a `PipelineRun` definition as `spec.pipelineSpec` or a single `Pipeline` defintion.

To make it work, migrations must be attached to corresponding task bundles as OCI artifacts,
typically with `oras attach`. Let's go deep dive a bit.

* If a task bundle has a migration, it must be annotated with `dev.konflux-ci.task.has-migration: true`
* A migration must be attached to the bundle with artifact type `text/x-shellscript` and annotated
  with `dev.konflux-ci.task.is-migration: true`.
* A task bundle has only one migration.
* A migration is written as a normal Bash script. Generally, it invokes `yq` to modify the pipelines.

Document [Task Migration]
(https://github.com/konflux-ci/build-definitions/?tab=readme-ov-file#task-migration) of

> konflux-ci/build-definitions describes the migrations in detailed. build-definitions provides rich

Reviewer: ```suggestion
konflux-ci/build-definitions describes the migrations in detail. build-definitions provides a rich
```

tool chain and CI for creation, validation and ensuring migrations are available to
pipeline-migration-tool.

## Commands

Pipeline migration tool supports several sub-commands to maintain the Konflux build pipelines.
`migrate` is the main one to do migrations. Following are a few examples of the commands usage. For
detailed information, refer to the help messages by running each one with `-h`.

### To apply migrations:

```bash
cd path/to/repo
pipeline-migration-tool migrate -u '<upgrades>'
```

`upgrades` is a JSON string encoded from a list of mappings. Each mapping includes data for a

> single task bundle upgrade, for example:

Reviewer: Is there some schema in the renovate docs that we could link to?

Or could we at least explain that these are Renovate upgrades?

> ```json
> [
>   {
>     "depName": "quay.io/konflux-ci/tekton-catalog/task-init",
>     "currentValue": "0.1",
>     "currentDigest": "sha256:...",
>     "newValue": "0.1",
>     "newDigest": "sha256:...",
>     "packageFile": ".tekton/component-name-pull.yaml",
>     "parentDir": ".tekton",
>     "depTypes": ["tekton-bundle"]
>   },
>   ...
> ]
> ```

Reviewer: nitpick: the three dots render with a red highlight (to show an error). Could be avoided like this I think:

```suggestion
```jsonc
[
  {
    "depName": "quay.io/konflux-ci/tekton-catalog/task-init",
    "currentValue": "0.1",
    "currentDigest": "sha256:...",
    "newValue": "0.1",
    "newDigest": "sha256:...",
    "packageFile": ".tekton/component-name-pull.yaml",
    "parentDir": ".tekton",
    "depTypes": ["tekton-bundle"]
  },
  // ...
]
```
```
